ml/src/main/scala/org/apache/spark/ml/spark/ProgressListener.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.ml.spark
+
+import org.apache.spark.scheduler._
+import org.apache.spark.storage.RDDInfo
+import water.Job
+
+/**
+  * One progress listener for each instance of model creation. Each listener gets all events so this particular one
+  * will respond only to the ones corresponding to a given RDD used for training (which should not be shared) and
+  * to a particular job type (depending on the algorithm).
+  *
+  * It might not be 100% accurate
+  */
+class ProgressListener(val job: Job[_], val rddInfo: RDDInfo, val stageNames: Iterable[String]) extends SparkListener {
+
+  override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+    if(jobStart.stageInfos.exists{ stage =>
```

Comment:
Hm not sure if I can change the total amount of work when the model is being built already, need to check.

Working with Spark's stages rather than jobs might be a bit confusing, though, here one job = one iteration of the algorithm that's why I can set the total amount of work to the number of iterations the user set in the UI and then just increment it by 1 with each new job.

It should be recorded on job start which job is running and incremented on job end instead but I was worried about this being run in parallel on many nodes and being harder to track but I can try.